api/v1alpha1/mariadb_webhook.go

```diff
@@ -94,23 +93,6 @@ func (r *MariaDB) ValidateDelete() (admission.Warnings, error) {
 }
 
 func (r *MariaDB) validateHA() error {
-	if r.Replication().Enabled && r.IsGaleraEnabled() {
```

Comment:
Could we leave the current validations as they are and make them configurable?

Although [CEL was introduced as alpha in 1.26](https://kubernetes.io/blog/2022/12/20/validating-admission-policies-alpha/), I would like to do a progressive migration instead of getting rid of our current validation straight away.

We would announce CEL support in our release notes for those who are in Kubernetes 1.26 or greater, and at the same time urge people in older versions to migrate, saying that the current webhook validations will be deprecated in further releases.

In any case, having duplicate validations is not ideal, I would make them configurable via the `WEBHOOK_LEGACY_VALIDATIONS_DISABLED` environment variable to be injected in the webhook deployment. Basically, whenever this env variable is present we will be skipping elements in the following slice:
- https://github.com/mariadb-operator/mariadb-operator/blob/b5836f17930f3010bd9313d25a949cdd574f6559/api/v1alpha1/mariadb_webhook.go#L45

What do you think?

The rationale behind using and env variable to control this instead of checking the Kubernetes versions is that the webhook is quite limited in regards to extensibility:
- https://github.com/mariadb-operator/mariadb-operator/blob/main/api/v1alpha1/mariadb_webhook.go

Since the `ValidateCreate` and `ValidateUpdate` methods need to be defined in the CR, there is no way we can have a single discovery client, instead we would need to instanciate a discovery client per each request, leading to increased memory allocations. We could use a singleton available through a shared variable, but I am not a big fan of this approach, it is more complex and I think it doesn't really worth it for this case.

Reply:
Because of the operator is on `v1alpha1` i thought i can be more "aggressive", but providing both checks with a deprecation phase is also a good idea. It means we need a Kubernetes version check [for the CEL rules](https://github.com/mariadb-operator/mariadb-operator/blob/8e5bbbc9eea4f5be96954796ed39a58dd052726a/api/v1alpha1/mariadb_types.go#L531-L536) to be able to still support older Kubernetes releases. On the other hand `1.26` is already [EOL](https://kubernetes.io/releases/)...

Reply:
Kubernetes `1.22` fails to import the CRD definitions if the validation code is included.
```
error: error validating "crds.yaml": error validating data: ValidationError(CustomResourceDefinition.spec.versions[0].schema.openAPIV3Schema.properties.spec): unknown field "x-kubernetes-validations" in io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1.JSONSchemaProps; if you choose to ignore these errors, turn validation off with --validate=false
```
This means we would need to generate two `crds.yaml` files (with and without the validation code) and we cannot use Helm to apply the correct one for a certain cluster, because Helm does not support the [templating of CRD definitions](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#some-caveats-and-explanations)...

Reply:
I'm planing to release `v0.0.30` shortly and the idea is to provide some visibility about our initial CEL support introduced in https://github.com/mariadb-operator/mariadb-operator/pull/783, also referring to the current PR to highlight that we are still working on it. I would prefer to roll this out progressively to see if people is actually facing other problems we haven't considered yet.

At the same time, I will be using this opportunity to say that our minimum supported version is .126, encouraging people to migrate. 

> Kubernetes 1.22 fails to import the CRD definitions if the validation code is included.

Thank you for testing this! I was testing the `release-v0.0.30` branch with Kubernetes 1.22, which has changes introduced in https://github.com/mariadb-operator/mariadb-operator/pull/783, and therefore also the `x-kubernetes-validations` field but I couldn't reproduce the problem. Maybe it has to do with something in the current branch?

```bash
❯ export KIND_IMAGE=kindest/node:v1.22.0

❯ make cluster-delete && make cluster
/home/martin/go/src/github.com/mariadb-operator/mariadb-operator/bin/kind delete cluster --name mdb
Deleting cluster "mdb" ...
Deleted nodes: ["mdb-control-plane"]
/home/martin/go/src/github.com/mariadb-operator/mariadb-operator/bin/kind create cluster --name mdb --image kindest/node:v1.22.0
Creating cluster "mdb" ...
 ✓ Ensuring node image (kindest/node:v1.22.0) 🖼
 ✓ Preparing nodes 📦
 ✓ Writing configuration 📜
 ✓ Starting control-plane 🕹️
 ✓ Installing CNI 🔌
 ✓ Installing StorageClass 💾
Set kubectl context to "kind-mdb"
You can now use your cluster with:

kubectl cluster-info --context kind-mdb

Not sure what to do next? 😅  Check out https://kind.sigs.k8s.io/docs/user/quick-start/

❯ make docker-build && make docker-load
docker buildx build -t docker-registry3.mariadb.com/mariadb-operator/mariadb-operator:v0.0.30 . --load
/home/martin/go/src/github.com/mariadb-operator/mariadb-operator/bin/kind load docker-image --name mdb docker-registry3.mariadb.com/mariadb-operator/mariadb-operator:v0.0.30
Image: "docker-registry3.mariadb.com/mariadb-operator/mariadb-operator:v0.0.30" with ID "sha256:3418ab5d5122df88db981808bba8bac6425571ddd773e09df75627bf0f994937" not yet present on node "mdb-control-plane", loading...

❯ helm install mariadb-operator deploy/charts/mariadb-operator
NAME: mariadb-operator
LAST DEPLOYED: Thu Aug 22 11:21:41 2024
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
mariadb-operator has been successfully deployed! 🦭

Not sure what to do next? 😅 Check out:
https://github.com/mariadb-operator/mariadb-operator#quickstart

❯ kubectl get pods
NAME                                                READY   STATUS    RESTARTS   AGE
mariadb-operator-558785774-clll9                    1/1     Running   0          49s
mariadb-operator-cert-controller-7c4b7c5dd5-48z9n   1/1     Running   0          49s
mariadb-operator-webhook-98765c9c-jmrcz             1/1     Running   0          49s
``` 

> On the other hand 1.26 is already [EOL](https://kubernetes.io/releases/)

Absolutely, it's EOL I get your point. However, in my experience, Kubernetes version pace is faster than people migrating. My suggestion is to include a notice in the release notes and offer users a little more time to make the transition.

Reply:
REAMDE.md updated:
- https://github.com/mariadb-operator/mariadb-operator/tree/release-v0.0.30?tab=readme-ov-file#kubernetes

Kubernetes v1.26 will be the minimum supported version starting from mariadb-operator v0.0.30. This is a step towards getting everyone migrated to a version with CEL, so we can continue working on this PR and roll it out in further versions.

Reply:
> v0.0.30 REAMDE.md updated:
> 
>     * https://github.com/mariadb-operator/mariadb-operator/tree/release-v0.0.30?tab=readme-ov-file#kubernetes
> 
> 
> Kubernetes v1.26 will be the minimum supported version starting from mariadb-operator v0.0.30. This is a step towards getting everyone migrated to a version with CEL, so we can continue working on this PR and roll it out in further versions.

That sounds good.